core/src/main/java/org/apache/calcite/rel/rules/PruneEmptyRules.java

```diff
@@ -540,4 +540,32 @@ public interface JoinRightEmptyRuleConfig extends PruneEmptyRule.Config {
       };
     }
   }
+
+
+  public static final RelOptRule EMPTY_TABLE =
+      ImmutableEmptyTableOptimizationConfig.of()
+          .withOperandSupplier(b0 -> {
+            return b0.operand(TableScan.class).noInputs();
+          })
+          .withDescription("ConvertEmptyTableToValues")
+          .toRule();
+
+  /** Configuration for rule that transforms an empty table into an empty values node.
+   * MaxRowCount is used as the stat to transform, hence Table implementer needs
+   * to supply this metadata if this optimization needs to be applied.*/
+  @Value.Immutable
+  public interface EmptyTableOptimizationConfig extends PruneEmptyRule.Config {
+
+    @Override default PruneEmptyRule toRule() {
+      return new PruneEmptyRule(this) {
+        @Override public void onMatch(RelOptRuleCall call) {
+          TableScan tableScan = call.rel(0);
+          Double maxRowCount = call.getMetadataQuery().getMaxRowCount(tableScan);
+          if (maxRowCount != null && maxRowCount == 0.0) {
+            call.transformTo(call.builder().push(tableScan).empty().build());
```

Comment:
The effect of this rule is very similar to what `RemoveEmptySingleRule` is doing. The latter is also trying to conserve the traits so maybe it is a good idea to attemp to conserve the traits as well here.